src/foam/net/node/CachingResponse.js

```diff
@@ -0,0 +1,119 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'foam.net.node',
+  name: 'CachingResponse',
+  extends: 'foam.net.node.CachedResponse',
+  implements: [ 'foam.net.node.ServerResponse' ],
+
+  documentation: `Response decorator that stores response data and produces
+      a CachedResponse when this response is finalized.`,
+
+  requires: [ 'foam.net.node.CachedResponse' ],
+  imports: [
+    'error',
+    'info'
+  ],
+
+  topics: [
+    'errored',
+    'recorded'
+  ],
+
+  properties: [
+    {
+      name: 'id',
+      factory: function() { return this.CachedResponse.ID_FROM_REQ(this.req); }
+    },
+    {
+      class: 'FObjectProperty',
+      of: 'foam.net.node.ServerRequest',
+      name: 'req',
+      documentation: 'Initial request to be cached.',
+      transient: true,
+      required: true,
+      final: true
+    },
+    {
+      class: 'FObjectProperty',
+      of: 'foam.net.node.ServerResponse',
+      name: 'res',
+      documentation: 'Initial response to be cached.',
+      transient: true,
+      required: true,
+      final: true
+    },
+    {
+      name: 'responseError',
+      transient: true,
+      value: null
+    }
+  ],
+
+  methods: [
+    function getStatusCode() {
+      return this.res.getStatusCode();
+    },
+    function setStatusCode(statusCode) {
+      if ( this.responseError ) return;
```

Comment:
Should these more aggressively throw errors? Are they expected to ever get called more than once?

The more-work-but-future-proof solution could be to copy-on-write if the cached response does get altered after retrieval.